### DISCUSSION TITLE
Is there a way to add context variables (or reach) in pymc

### DISCUSSION THREAD

**Discussion by shuvayan** [pymc-labs/pymc-marketing#691] [Q&A]
Hello,

I would like to know if there is a way to add variables (like impressions) which capture the exposure of the advertising spends in the pymc model structure. 
Similar to what gets inputted to Robyn.
Also , if we think about it , Spend Creates Reach via Impression and thereafter comes Revenue. So it should be something of a spend, reach and sales response hyperplane which has a simplified representation in the sales response curve.

I might be completely off the track here so please feel free to reject the idea/assumption.
Also, if there a way to generate Robyn like one-pagers( Clients seem to love those, no matter how bad penalized regression can be for MMM). :)

Note: This is not an issue per say, but more of a discussion point

**Comment by williambdean** [comment]
Hi @shuvayan 

So I am reading two items here: 1. Adding impressions into the MMM and 2. one-pagers. Please correct me if I am wrong

1. Would you be able to using impressions as the media features instead of spends? What do you have in mind in order to incorporate both media impressions and media spend?

2. I am not too sure what this is so feel free to elaborate. If it is some type of presentation of the results, it seems like it is out of the scope the repo imo. But it there are missing analysis or technical elements derived from the model that work well with clients, it would be great to hear what works in your experience!

**Reply by thipokKub** [reply]
Hello, I was thinking about alternative formulation to model marketing impression / reach / convert dynamics. Currently it just a sketch, but would it make sense to model the marketing result as a compartmental model? It could be that the saturation effect is the result of the following

Let's say that ads platform like google / meta start out by ranking what cohort of people to best response for a given campaign, then randomly select audience from this pool. Thus likelihood of responding to the campaign decrease over time. This result in the saturation curve. And it will take some time for the ads platform to identify quality audience

As for carryover effect, let's say that not every people react to the social media immediately i.e. the transition rate to different state is lower than the combined output rate. It is like filling a leaking water bucket once, but the leaking effect is slow. So it took some time to drain out. Therefore for each cohort, they should have a characteristics of different response time. However, once they react, they were removed from the ads potential pool. Lastly, the can model Ads Spending as an add-on effect, if we already has a base conversion rate

In this way we could incorporate the funneling effect, and estimate the dynamics e.g. cohort > impression > reach > convert. We can then fitted the model parameter and solve for the ODE. We can then adjust the funnel dynamics for different platform depends on their provided information

The limitation of this model is we assume that the characteristic of each do not change i.e. we don't target new group, so the response rate is relatively stable. Moreover, we assume no interaction between different channel (no competing, and combos), and the initial pool size is fixed (unobserved?)

The state transition is something like

<img width="904" alt="image" src="https://github.com/pymc-labs/pymc-marketing/assets/21255489/0a8c91a9-0aed-4980-8525-a6b66bb430fc">

or

<img width="655" alt="image" src="https://github.com/pymc-labs/pymc-marketing/assets/21255489/2bc281bf-56cf-4362-a4ec-6ad626567424">

I think with some specific formulation, the Impression + Reach stage can be merge into a single stage

However, if we assume that the ads platform will rank potential audience first, then expose the selected pool through out the campaign lifetime. The quality of audience will reduce overtime. So we can expect the transition rate to reduce for impress / reach / convert / interact, and increased rate for exhaust / dead. We can mimic this effect by gradually change the rate overtime, by decay, and / or saturation

Finally, There seems to be some literature, but not directly for ads platform. So I'm not sure this should be the way or not 🤔

**Note**
- If we assume that ads platform ranking model is a binary classification, and a fixed pool size. Given the estimated parameters, we should be able to determine model performance through GINI coefficient. When assume parametric GINI curve, by measuring how the performance degrade overtime
- I've just realized that using all the first letter of each stage, it can be arrange into *P.R.I.C.E.D*

---

Alternatively, there is a [Hellofresh's blog about MMM](https://engineering.hellofresh.com/bayesian-media-mix-modeling-using-pymc3-for-fun-and-profit-2bd4667504e6) which fit the model in multiple stage, and adjust for the funneling. Albeit their funneling is more about ads channel tier, but it should be applicable to impression / reach / convert staging. Another one I've found is [this technical report](https://arxiv.org/pdf/2311.05587v5), but IMO, it stretch a lot of concept a bit too much

**Comment by shuvayan** [comment]
Hello @wd60622 , @thipokKub ,

Thank you both for creating the wonderful discussion thread. 
I think there is meat in this approach as come to think of it, media spend conversion happens through impressions.

2. I spend $100 , my ad reaches 10 people out of which 2 sign up on my page. It is simplistic example but demonstrates the flow of effect and cause from spend -> impressions -> joins.
I think google meridian is somewhat capturing this using the reach parameter in their objective function , but not sure.

Overall , it would be ideal if such cases can be dealt with something like a nested model within the bayesian framework, but not sure if that is statistically possible.

2.About the one pager , below is a sample output for reference:

![Screenshot 2024-05-22 at 9 39 37 AM](https://github.com/pymc-labs/pymc-marketing/assets/11851253/d55d5a01-f967-4f58-bcad-debff4585957)


The in cluster bootstrapped ROI is not relevant for pymc , so we can ignore it. 
I think we should be able to recreate the remaining charts for pymc too , as clients like to see all these metrics together and having such a one pager really helps from the presentation perspective.

Do let me know if these make sense.

**Comment by jadenguyen-posti** [comment]
Hello,

I've just come across this interesting discussion here and have a slightly related question. What's your experience building the model using different marketing metrics spend / impression / reach / conversion? 

I've tried building two models using the same dataset but two marketing metrics - spend and impression. I noticed a small difference in accuracy between them (impression data generated better accuracy), although both models produced quite similar results when it came to channel contribution to sales.

I think lower-funnel metrics (reach, conversion) are generally more relevant for estimating marketing effectiveness and contribution to business KPIs like sales. But it brings its own challenge that not all marketing activities and campaigns are designed to drive direct sales and data like conversion isn't always available. 

So I'm curious, does it really matter which metric we choose, given these metrics tend to be correlated? Would like to hear your experience and opinions on this topic :)
